pkg/pillar/netdump/netdump.go

```diff
@@ -0,0 +1,375 @@
+
+// LastPublishAt returns timestamp of the last publication for any of the given topics.
+// Returns zero timestamp (and nil error) if nothing has been published yet.
+func (dumper *NetDumper) LastPublishAt(topics ...string) (time.Time, error) {
+	var lastPub time.Time
+	// ReadDir returns files sorted in the ascending order
+	// (with our filename layout this means starting with the oldest).
+	dirEntries, err := os.ReadDir(netdumpDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return lastPub, nil
+		}
+		return lastPub, fmt.Errorf("netdump: failed to list directory %s: %w",
+			netdumpDir, err)
+	}
+	for _, topic := range topics {
+		netdumpPrefix := topic + "-"
+		for i := len(dirEntries) - 1; i >= 0; i-- {
+			filename := dirEntries[i].Name()
+			if !strings.HasSuffix(filename, netdumpExtension) ||
+				!strings.HasPrefix(filename, netdumpPrefix) {
+				continue
+			}
+			timeStr := strings.TrimPrefix(filename, netdumpPrefix)
+			timeStr = strings.TrimSuffix(timeStr, netdumpExtension)
+			timestamp, err := time.ParseInLocation(netdumpTimestampFormat, timeStr, time.UTC)
+			if err != nil {
+				return time.Time{}, fmt.Errorf("netdump: failed to parse timestamp %s: %w",
+					timeStr, err)
+			}
+			if lastPub.IsZero() || timestamp.After(lastPub) {
+				lastPub = timestamp
+			}
+			break // next topic
+		}
+	}
+	return lastPub, nil
+}
+
+// Publish a single netdump archive, containing information such as the current
+// DPCL (DevicePortConfigList), DNS (DeviceNetworkStatus), ifconfig output, DHCP
+// leases, wwan status and more.
+// Additionally, it is possible to attach traces and packet captures recorded
+// during an execution of one or more network requests (i.e. requests targeted
+// at remote endpoints and carried over the network).
+// In order to avoid race conditions between microservices, each microservice
+// should use different topic(s) (e.g. topic = microservice name).
+// Topic name should not contain characters which are not suitable for filenames
+// (across all OSes).
+func (dumper *NetDumper) Publish(topic string,
+	requests ...TracedNetRequest) (filepath string, err error) {
+
+	// Generate name for the network dump.
+	timestamp := time.Now().UTC()
+	netdumpPrefix := topic + "-"
+	netdumpName := netdumpPrefix + timestamp.Format(netdumpTimestampFormat)
+	filepath = path.Join(netdumpDir, netdumpName+netdumpExtension)
+
+	// Create directories inside the archive.
+	files := []fileForTar{
+		{
+			dstPath: netdumpName,
+			isDir:   true,
+		},
+		{
+			dstPath: path.Join(netdumpName, "eve"),
+			isDir:   true,
+		},
+		{
+			dstPath: path.Join(netdumpName, "linux"),
+			isDir:   true,
+		},
+	}
+	if len(requests) > 0 {
+		files = append(files, fileForTar{
+			dstPath: path.Join(netdumpName, "requests"),
+			isDir:   true,
+		})
+	}
+
+	// Include some files written by EVE or EVE installer that contain
+	// information useful for network troubleshooting.
+	for _, file := range eveFilesToDump {
```

Comment:
does it make sense to separate the eve and linux files (since they are most likely the same) from the different nettrace coonection items? I would think each nettrace connection can have a dump file, then add a system eve/linux/routing/etc related dump file for the same collection time. Otherwise, when adding more nettrace connections, all those eve/linux items would be replicated every time.

Reply:
When nim, zedagent or downloader will for example try multiple source IP addresses and interfaces, traces from all attempts will actually end up in the same netdump (notice that `NetDumper.Publish` takes variable arrays of traced requests). So there are not going to be netdumps close to each other time-wise that it would make sense to capture the EVE and Linux state only once.
Secondly, I wanted a single netdump tarball to be self-contained and (hopefully) provide all information needed for issue troubleshooting. Although customers will be able to package all netdumps with one `edgeview tar` call, I do not want it to be a requirement.
Btw. the contribution these files make to the final tarball size is rather minimal, only ~9KB.